src/ucp/core/ucp_worker.c

```diff
@@ -2345,31 +2371,62 @@ static unsigned ucp_worker_discard_uct_ep_progress(void *arg)
     return 1;
 }
 
-static int ucp_worker_discard_remove_filter(const ucs_callbackq_elem_t *elem,
-                                            void *arg)
+static int
+ucp_worker_discard_remove_filter(const ucs_callbackq_elem_t *elem, void *arg)
 {
-    if ((elem->cb != ucp_worker_discard_uct_ep_destroy_progress) &&
-        (elem->cb != ucp_worker_discard_uct_ep_progress)) {
+    if ((elem->arg != arg) ||
+        ((elem->cb != ucp_worker_discard_uct_ep_destroy_progress) &&
+         (elem->cb != ucp_worker_discard_uct_ep_progress))) {
         return 0;
     }
 
-    ucp_worker_put_flush_req((ucp_request_t*)elem->arg);
+    ucp_worker_discard_uct_ep_complete((ucp_request_t*)elem->arg);
     return 1;
 }
 
-static void ucp_worker_discarded_uct_eps_cleanup(ucp_worker_h worker)
+static void
+ucp_worker_discard_uct_ep_pending_count_purged(uct_pending_req_t *self,
+                                               void *arg)
 {
+    ++(*((unsigned*)arg));
+}
+
+static void ucp_worker_discard_uct_ep_cleanup(ucp_worker_h worker)
+{
+    ucp_request_t *req;
+    unsigned num_purged;
     uct_ep_h uct_ep;
 
-    ucs_callbackq_remove_if(&worker->uct->progress_q,
-                            ucp_worker_discard_remove_filter, NULL);
+    /* Destroying UCP EP in ucp_ep_disconnected() could start UCT EP discarding
+     * operations (e.g. for TMP EP in CM lane). Do cleanup of discarding
+     * functionality after trying to destroy UCP EPs in order to destroy all
+     * remaining UCP EPs here */
 
-    /* If ep owns the discard operation ep_destroy will cancel it. We are after
-     * uct_worker_progress_unregister_safe and ucp_worker_discard_remove_filter,
-     * so either we canceled req or it was finished and removed from kh before
-     */
-    kh_foreach_key(&worker->discard_uct_ep_hash, uct_ep, {
+    kh_foreach(&worker->discard_uct_ep_hash, uct_ep, req, {
+        ucs_assert(uct_ep == req->send.discard_uct_ep.uct_ep);
+
+        num_purged = 0;
+        uct_ep     = req->send.discard_uct_ep.uct_ep;
+
+        uct_ep_pending_purge(uct_ep,
+                             ucp_worker_discard_uct_ep_pending_count_purged,
```

Comment:
maybe just do ucp_worker_discard_uct_ep_check_no_progress+ucp_worker_discard_uct_ep_complete from the purge callback?

Reply:
done

Reply:
seems not fully done - no need to count anything in the purge callback, and need to rename the purge callback

Reply:
done